cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -97,6 +103,46 @@ public static void validateConfig(SecureStore secureStore, SourceControlConfig s
     }
   }
 
+  /**
+   * Returns the <a href="https://git-scm.com/docs/git-hash-object">Git Hash</a>
+   * of the requested file path in the current HEAD. For symlinks, it returns the hash of
+   * the target file. This ensures that the file hash of a symlink is equal to the hash of the target file. It
+   * doesn't return the hash for directories. If the directory contains symlinks, we can't depend on the
+   * directory hash to ensure the contents are unchanged.
+   *
+   * @param path The absolute path on the filesystem that should include {@link RepositoryManager#getBasePath()} as a
+   *             prefix.
+   * @return The git file hash of the requested file.
+   * @throws IOException              when file isn't found, there are circular symlinks or other file IO errors.
+   * @throws IllegalArgumentException when the provided file isn't a regular file, eg: directory.
+   * @throws NotFoundException        when the file isn't committed to Git.
+   * @throws IllegalStateException    when {@link RepositoryManager#startUp()} isn't called before this.
+   */
+  public String getFileHash(Path path) throws IOException, NotFoundException {
+    validateInitialized();
+    // Git considers symlinks as separate files. So the hash of the symlink will be different from the target file.
+    // Resolve path in case of symlinks. This ensures the hash of the target file is returned.
+    path = path.toRealPath();
+    // We should not be comparing hashes of directories. If the directory contains symlinks, we can't depend on the
+    // directory hash to ensure the contents are unchanged.
+    if (!Files.isRegularFile(path)) {
+      throw new IllegalArgumentException(String.format("Path %s doesn't refer to a regular file.",
+                                                       path.toAbsolutePath()));
+    }
+    Path relativePath = sourceControlConfig.getLocalRepoPath().relativize(path);
+    // HEAD refers to the checked out commit.
+    ObjectId lastCommitId = git.getRepository().resolve(Constants.HEAD);
+    // Each commit points to a tree that contains the files/directories as nodes. The ObjectId (hashes) of the nodes
+    // represent the state of a file at that commit.
+    RevCommit commit = new RevWalk(git.getRepository()).parseCommit(lastCommitId);
+    // Find the node representing the exact file path in the tree.
+    TreeWalk walk = TreeWalk.forPath(git.getRepository(), relativePath.toString(), commit.getTree());
+    if (walk == null) {
+      throw new NotFoundException("File not found in Git revision tree.");
+    }
+    return walk.getObjectId(0).getName();
```

Comment:
So, are we returning a file hash or a commit hash? The function name and description seems like talking about file hash, but implementation looks more like commit hash

Reply:
IIUC `git.getRepository().resolve(Constants.HEAD)` will get the commit ID. This will be a node that is the root of the directory at that particular commit. We then need to traverse the tree to reach the file node we require. This is done by  `TreeWalk.forPath(git.getRepository(), relativePath.toString(), commit.getTree())`. 
I'm confident that the hash is actually the file hash because I have a test case for it. If you check `RepositoryManaagerTest#testGetFileHashSuccess`,  I've found the git file hash using the following command and compared it with the one returned by my code.
```
$ printf "{name: pipeline}" > ./pipeline.json && git hash-object ./pipeline.json
5905258bb958ceda80b6a37938050ad876920f09
```

Reply:
yep, `git hash-object` actually computes the sha1 hash by `prefixing the object with "blob ", followed by the length (as a human-readable integer), followed by a NUL character`: https://alblue.bandlem.com/2011/08/git-tip-of-week-objects.html

fact: 
```
$ printf 'blob 16\0{name: pipeline}' | shasum
5905258bb958ceda80b6a37938050ad876920f09  -
```

Reply:
Let's use this approach (doing sha) in tests, it will ensure it does not break and will clarify the code

Reply:
Tests now calculate sha1 hashes.